From 8f9aa67d52a4b128da44947d10ee0dca5c165a5e Mon Sep 17 00:00:00 2001
From: Keyon Jie <yang.jie@linux.intel.com>
Date: Sun, 28 Apr 2019 18:06:52 +0800
Subject: [PATCH] FROMGIT: ASoC: SOF: topology: refine multi-core power up/down to avoid
 core status chaos

BugLink: https://bugs.launchpad.net/bugs/1826181

We should not power up DSP core when it is already on, vice versa, don't
power down it when it is already off, here add checks to fix it.

Todo: add ref_cnt for each DSP cores to manage the power status of those
cores.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>
(cherry picked from commit 81160d3f7d6ecf78cb20830ba80690c1d4da5330
git://github.com/thesofproject/linux.git)
Signed-off-by: Hui Wang <hui.wang@canonical.com>
---
 sound/soc/sof/topology.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/sound/soc/sof/topology.c b/sound/soc/sof/topology.c
index 432ae343f960..2c1aa0cca2b3 100644
--- a/sound/soc/sof/topology.c
+++ b/sound/soc/sof/topology.c
@@ -1358,6 +1358,9 @@ int sof_load_pipeline_ipc(struct snd_sof_dev *sdev,
 		return ret;
 	}
 
+	if (sdev->enabled_cores_mask & (1 << pipeline->core))
+		return ret;/* core already enabled, return */
+
 	/* power up the core that this pipeline is scheduled on */
 	ret = snd_sof_dsp_core_power_up(sdev, 1 << pipeline->core);
 	if (ret < 0) {
@@ -2146,9 +2149,19 @@ static int sof_widget_unload(struct snd_soc_component *scomp,
 		}
 		break;
 	case snd_soc_dapm_scheduler:
+		pipeline = swidget->private;
+
+		/*
+		 * For core 0, we will power down it at PM ops.
+		 * For other cores, if it is already powered off, just return.
+		 * TODO: add ref counts for each core, only power off it when
+		 *	 ref counts is decreased by 0.
+		 */
+		if (!(sdev->enabled_cores_mask & (1 << pipeline->core)) ||
+		    !pipeline->core)
+			return 0;
 
 		/* power down the pipeline schedule core */
-		pipeline = swidget->private;
 		ret = snd_sof_dsp_core_power_down(sdev, 1 << pipeline->core);
 		if (ret < 0)
 			dev_err(sdev->dev, "error: powering down pipeline schedule core %d\n",
-- 
2.17.1

